File: src/micro_lqr_controller/tools/design_lqr.m
%% Standard discrete LQR design for the fixed-leg two-wheel robot
% State order: x = [pitch; pitch_rate; position; velocity]
% Model input: TOTAL axle torque, u_total = tau_left_forward + tau_right_forward
% The ROS 2 node builds this same model and solves the DARE at startup.

clear; clc;

%% Physical parameters -- replace with measured/CAD values
m = 2.54;                 % body mass participating in pitch [kg]
M = 0.26;                 % non-pitch translating mass [kg]
Jw = 0.0;                 % each wheel rotational inertia [kg*m^2]
h = 0.120;                % axle-to-body-COM distance [m]
I = 0.036576;             % body pitch inertia about COM [kg*m^2]
r = 0.030;                % wheel radius [m]
g = 9.80665;              % gravity [m/s^2]
Ts = 0.003;                % controller period [s]
use_course_legacy_b4 = false;

% Equivalent translational mass includes two wheel rotational inertias.
Meq = M + 2*Jw/r^2;
D = (Meq+m)*(m*h^2 + I) - m^2*h^2;

A = [0, 1, 0, 0;
     ((Meq+m)*m*g*h)/D, 0, 0, 0;
     0, 0, 0, 1;
     -(m^2*g*h^2)/D, 0, 0, 0];

if use_course_legacy_b4
    % Exact fourth B entry found in the uploaded course wheel_control.m.
    b4 = 1/((Meq+m)*r) - (m^2*h^2)/((Meq+m)*D*r);
else
    % Standard cart-pole result after solving the coupled equations.
    b4 = (m*h^2 + I)/(D*r);
end
B = [0; -(m*h)/(D*r); 0; b4];
C = eye(4);
Dmat = zeros(4,1);

%% Cost settings
Q = diag([1, 1, 1, 1]);
R = 10;                   % course example uses 1; 10 is safer for first hardware test

%% Exact zero-order-hold discretization and discrete LQR
sys_c = ss(A, B, C, Dmat);
sys_d = c2d(sys_c, Ts, 'zoh');
Ad = sys_d.A;
Bd = sys_d.B;
[K, P, poles] = dlqr(Ad, Bd, Q, R);

fprintf('State order: [pitch pitch_rate x x_dot]\n');
fprintf('Input: total axle torque [N*m]\n');
fprintf('K = [%.12g, %.12g, %.12g, %.12g]\n', K(1), K(2), K(3), K(4));
fprintf('max(abs(closed-loop poles)) = %.12g\n', max(abs(poles)));
fprintf('controllability rank = %d\n', rank(ctrb(Ad, Bd)));

fprintf('\nYAML manual-gain comparison block:\n');
fprintf('lqr.use_manual_gain: true\n');
fprintf('lqr.manual_k_pitch: %.12g\n', K(1));
fprintf('lqr.manual_k_pitch_rate: %.12g\n', K(2));
fprintf('lqr.manual_k_position: %.12g\n', K(3));
fprintf('lqr.manual_k_velocity: %.12g\n', K(4));

%% Optional linear closed-loop initial-condition simulation
x0 = [3*pi/180; 0; 0; 0];
N = round(5/Ts);
x = zeros(4, N+1);
u = zeros(1, N);
x(:,1) = x0;
for k = 1:N
    u(k) = -K*x(:,k);
    x(:,k+1) = Ad*x(:,k) + Bd*u(k);
end
t = (0:N)*Ts;
figure('Name','Discrete LQR linear model');
subplot(3,1,1); plot(t, x(1,:)*180/pi); grid on; ylabel('pitch [deg]');
subplot(3,1,2); plot(t, x(3,:)); grid on; ylabel('x [m]');
subplot(3,1,3); stairs(t(1:end-1), u); grid on; ylabel('u total [N m]'); xlabel('time [s]');

File: src/micro_lqr_controller/tools/design_lqr.py
#!/usr/bin/env python3
"""Reproduce the ROS 2 node's exact-ZOH discrete LQR design."""
from __future__ import annotations

import argparse
import numpy as np
from scipy.linalg import expm, solve_discrete_are, eigvals


def design(
    body_mass: float = 2.54,
    non_pitch_mass: float = 0.26,
    wheel_inertia_each: float = 0.0,
    com_height: float = 0.120,
    body_pitch_inertia: float = 0.036576,
    wheel_radius: float = 0.030,
    gravity: float = 9.80665,
    dt: float = 0.003,
    q=(1.0, 1.0, 1.0, 1.0),
    r_cost: float = 10.0,
    course_legacy_b4: bool = False,
):
    m = body_mass
    M = non_pitch_mass + 2.0 * wheel_inertia_each / wheel_radius**2
    h = com_height
    I = body_pitch_inertia
    den = (M + m) * (m * h**2 + I) - m**2 * h**2
    if den <= 0:
        raise ValueError("Model denominator is non-positive")

    A = np.array(
        [
            [0.0, 1.0, 0.0, 0.0],
            [((M + m) * m * gravity * h) / den, 0.0, 0.0, 0.0],
            [0.0, 0.0, 0.0, 1.0],
            [-(m**2 * gravity * h**2) / den, 0.0, 0.0, 0.0],
        ]
    )
    if course_legacy_b4:
        b4 = 1.0 / ((M + m) * wheel_radius) - (
            m**2 * h**2 / ((M + m) * den * wheel_radius)
        )
    else:
        b4 = (m * h**2 + I) / (den * wheel_radius)
    B = np.array([[0.0], [-(m * h) / (den * wheel_radius)], [0.0], [b4]])

    augmented = np.zeros((5, 5))
    augmented[:4, :4] = A
    augmented[:4, 4:] = B
    exact = expm(augmented * dt)
    Ad = exact[:4, :4]
    Bd = exact[:4, 4:]
    Q = np.diag(q)
    R = np.array([[r_cost]])
    P = solve_discrete_are(Ad, Bd, Q, R)
    K = np.linalg.solve(Bd.T @ P @ Bd + R, Bd.T @ P @ Ad)
    poles = eigvals(Ad - Bd @ K)
    ctrb = np.column_stack([Bd, Ad @ Bd, Ad @ Ad @ Bd, Ad @ Ad @ Ad @ Bd])
    return A, B, Ad, Bd, K, poles, np.linalg.matrix_rank(ctrb)


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--course-legacy-b4", action="store_true")
    parser.add_argument("--r", type=float, default=10.0)
    args = parser.parse_args()
    _, _, _, _, K, poles, rank = design(
        r_cost=args.r, course_legacy_b4=args.course_legacy_b4
    )
    print("state order: [pitch, pitch_rate, x, x_dot]")
    print("model input: total axle torque [N*m]")
    print("K:", np.array2string(K, precision=12))
    print("closed-loop poles:", np.array2string(poles, precision=12))
    print("max pole magnitude:", float(np.max(np.abs(poles))))
    print("controllability rank:", rank)


if __name__ == "__main__":
    main()
